### DISCUSSION TITLE
What is the correct way to handle errors inside getServerSideProps?

### DISCUSSION THREAD

**Discussion by lumenwrites** [vercel/next.js#40565] [Help]
Hi! Let's say I have code like this:
```javascript
export const getServerSideProps = async (context) => {
  try {
    functionThatMightFailAndThrowAnError()
  } catch (error) { 
    // What should I do here?
  }
  return { props: {} }
}
```

What should I do when the function throws an error? I want to log the error to my logging provider, and then show the user a 500 error page, but no matter what I do, Next doesn't seem to show me the 500 page.

Generally speaking, I'd like to understand the correct approach to error handling inside getServerSideProps. How do experienced developers on big projects usually do that?

**Comment by icyJoseph** [comment]
One naive approach is to return a props, which contains an error message, and from inside your component, you read if that's present in which case, you render some UI that helps your users understand what has happened:

```tsx
export const getServerSideProps = async (context) => {
  try {
    functionThatMightFailAndThrowAnError()
  } catch (error) { 
    await logger();
     return {props: {error: 'foo'}}
    // What should I do here?
  }
  return { props: {data} }
}

function MyComponent({data, error}) {
   if(error) return <ErrorComponent error={error} />

   return <Data  data={data} />
}
```

If you want to use the `next.js` built-ins, 400, 500, or customise with `_error`:

- `404` just return `{ notFound: true}``
- `500`/`_error`, log and then re-throw `throw error`... and test it on Build mode! on Dev you'll only see the Error Overlay (`500.tsx`, takes precedence over `_error.tsx` IIRC)

Note that an error on a page such as 500, that in turn crashes out the rendering process, loops back to a 500, which would crash out again, and loop back, and well, you know, it'll run away to infinity. Make sure your 500/error pages are solid and robust.
